From: Milan Crha <mcrha@redhat.com>
Date: Thu, 21 Aug 2025 12:48:34 +0200
Subject: cursor-renderer/native: Sanity check texture size before allocating

The GSettings' cursor size can be set to zero by some desktop environments,
like XFCE, then starting the Wayland session segfaults due to invalid
cursor size.

Closes https://gitlab.gnome.org/GNOME/mutter/-/issues/3933

Part-of: <https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/4592>

Bug-Ubuntu: https://bugs.launchpad.net/bugs/2125552
Origin: Upstream commit c7b64cd91992e6216c296512b1bd93d33dc0b726
---
 src/backends/native/meta-cursor-renderer-native.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/src/backends/native/meta-cursor-renderer-native.c b/src/backends/native/meta-cursor-renderer-native.c
index 351646f..949f922 100644
--- a/src/backends/native/meta-cursor-renderer-native.c
+++ b/src/backends/native/meta-cursor-renderer-native.c
@@ -801,6 +801,14 @@ scale_and_transform_cursor_sprite_cpu (MetaCursorRendererNative *cursor_renderer
   if (!src_texture)
     return NULL;
 
+  if (dst_width < 1 || dst_height < 1)
+    {
+      g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED,
+                   "Invalid size for cursor texture %d x %d",
+                   dst_width, dst_height);
+      return NULL;
+    }
+
   dst_texture = cogl_texture_2d_new_with_format (cogl_context,
                                                  dst_width,
                                                  dst_height,
